### PR TITLE
fix: bugs found in Hedera > Ethereum conversion framework

### LINKED ISSUE #11730: Run ConcurrentSuites.ethereumSuites
Run the tests in `ConcurrentSuites.ethereumSuites` and see if there are any problems. If there are a lot - define a new work package. If not - fix those errors

### LINKED ISSUE #12364: Change fuzzy matching logic to run _Eth tests
When we run a test with fuzzy matching and _Eth prefix we are not finding the snapshot file. We need to remove the _Eth suffix from the file name when we are searching for the snapshot file.

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:
Running all the `ConcurrentSuites.ethereumSuites` tests and fix all the issues related to the framework the converts Hedera to Ethereum contract calls and contract creates. The bugs found on service level were fixed in https://github.com/hashgraph/hedera-services/pull/12834

**Related issue(s)**:

Fixes #11730
Fixes #12364

**Notes for reviewer**:

**Solved problems**
* Update contract bytecode file, if we have constructor args: 
  * When we use HAPI transactions to create a contract that needs constructor arguments, ContractCreateTransactionBody has separate fields for initcodeSource/fileID and constructorParameters. So if args are needed the EVM payload is constructed (args are appended to the init code) in `com.hedera.node.app.service.contract.impl.infra.HevmTransactionFactory#initcodeFor` But when we use Ethereum transaction we can pass only CallData, which is either the bytecode or the fileID. There is no field in the transaction body to map the arguments. This is why when we convert HAPI to Eht creation and detect arguments, we perform additional file update and append the arguments to the bytecode.
* Try to convert args/params addresses to EVM addresses: 
  * When we try to access any mutable accounts from the EVM frame, we can do it only by EVM address (priority address). So when we convert to ethereum call/create, we also try to convertny params/args that are representing addresses from a long zero address to an EVM address, when it is possible. In order to achieve this, when create is done, we added a getContractInfo query and saved the EVM address of the contract to the spec registry. So when we map calls with params/args we iterate over them and check if they are addresses and if we have already saved the corresponding EVM one.

**Refused conversion**
* Skip ethereum contract creation conversion (assert admin key/renew account/memo/staking fields): 
  * EthereumTransactionBody doesn’t support these fields, so assertions against them will fail. 
* Refuse eth conversion because we can't set invalid bytecode to callData in ethereum transaction: 
  * When we perform HAPI create transactions, byte codes are uploaded as a file in separate transactions, and parsed as call data (payload) on the node itself. So any invalid byte code will result in some kind of error status. While in ethereum creates the byte code, it must be parsed as call data on the client side, and if it is invalid, we fail to send the transaction. 
* Refusing ethereum create conversion, because we get INVALID_SIGNATURE upon tokenAssociate, since we have CONTRACT_ID key: 
  * The default key validator doesn't allow as to sing tokenAssociate transactions with simple Contrac_ID keys.
* Refuse conversion because of MODIFYING_IMMUTABLE_CONTRAC: 
  * Since it is impossible to set an admin key, or any other special keys by ethereum create, our contracts become immutable. Any test that tries to modify it will fail.
* Refuse because of `__AddressBook.sol:AddressBook___________` placeholder
   * it would be overkill to support that for a single test

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hapi-utils/src/main/java/com/hedera/node/app/hapi/utils/ethereum/EthTxData.java [hunk 1]
```diff
@@ -402,6 +402,10 @@ private static EthTxData populateEip1559EthTxData(RLPItem rlpItem, byte[] rawTx)
             return null;
         }
 
+        final var value = rlpList.get(6).asBytes().length == 0
+                ? BigInteger.ZERO
+                : new BigInteger(rlpList.get(6).asBytes());
```
Comment (vtronkov) [inline hedera-node/hapi-utils/src/main/java/com/hedera/node/app/hapi/utils/ethereum/EthTxData.java:407]: This is a better way to handle two's complement conversion. Before when we were passing a negative number(for example -1000) the result was 271474976710656. If there is another way to handle this, please let me know and I will change it.

File: hedera-node/test-clients/src/main/resource/contract/contracts/MintContract/MintContract.bin [hunk 2]
```diff
@@ -1 +1 @@
...50509250929050565b67ffffffffffffffff8116811461088857600080fd5b60006020828403121561096a57600080fd5b81356108ed81610942565b60005b83811015610990578181015183820152602001610978565b838111156101355750506000910152565b600082516109b3818460208701610975565b9190910192915050565b8051600381900b811461087457600080fd5b6000602082840312156109e157600080fd5b6108ed826109bd565b634e487b7160e01b600052603260045260246000fd5b6000606082016001600160a01b0386168352602067ffffffffffffffff8616818501526060604085015281855180845260808601915060808160051b870101935082870160005b82811015610a8d57878603607f1901845281518051808852610a6e81888a01898501610975565b601f01601f191696909601850195509284019290840190600101610a47565b50939998505050505050505050565b600080600060608486031215610ab157600080fd5b610aba846109bd565b9250602080850151610acb81610942565b604086015190935067ffffffffffffffff811115610ae857600080fd5b8501601f81018713610af957600080fd5b8051610b0761075982610713565b81815260059190911b82018301908381019089831115610b2657600080fd5b928401925b82841015610b4d578351610b3e81610879565b82529284019290840190610b2b565b8095505050505050925092509256fea2646970667358221220539021dabefb3ad6b4b014421b748d2f5430fab6b0b0f31c72a38f86f0eabb5d64736f6c63430008090033
```
Comment (vtronkov) [inline hedera-node/test-clients/src/main/resource/contract/contracts/MintContract/MintContract.bin:1]: The change here is that I've removed **0x** at the beginning of the file. The reason for that is when we call `EthereumCallDataHydration.tryToHydrate` we try to decode the hex `Hex.decode(callDataFile.contents().toByteArray())` but the **0x** part is not hex. I saw that in most of our **.bin** files we don't have that so I think it should be removed.

I have two questions:
1. Should I remove the 0x prefix for all *.bin files we have that(`AssociateDissociate.bin, Associator.bin, MixedMintToken.bin, and NastedAssociateDissociate.bin`)?
2. In `EthereumCallDataHydration.tryToHydrate` when we try to decode the hex `Hex.decode(callDataFile.contents().toByteArray())` if there is an exception we throw `INVALID_FILE_ID`. Do you guys think there is a more suitable error we can throw there? It's a bit misleading, at least in my case.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java [hunk 3]
```diff
@@ -731,6 +731,8 @@ final HapiSpec cannotUseMirrorAddressOfAliasedContractInPrecompileMethod() {
                 .then();
     }
 
+    // refusingEthConversion because the minters contract has placeholders that the HapiEthereumContractCreate doesn't
+    // support
```
Comment (vtronkov) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java:735]: The `Minters` contract has a placeholder(`__AddressBook.sol:AddressBook___________`) that we are replacing in `bookInterpolated` before uploading the contract. When we do `HapiContractCreate` we don't use the bytecode, instead, we just pass the fileId(and later on we get the bytecode itself). For the `EthereumContractCreate` we do need the bytecode so we can construct `EthTxData`. In order to do that we have two options, as far as I can see:
1. Refactor the `EthereumContractCreate` and `EthTxData` to use the fileId
2. Give `EthereumContractCreate` access to the `HapiSpecOperations` and get the `HapiFileUpdate` by name that is replacing the empty byte contents(from create) with the corrected placeholder replaced bytecode(the logic in `createLargeFile`).
Both of these seem like an overkill to support only for a single test that we have.
There might be some more elegant way that I do not see. If you have an idea I can try it :)

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java [hunk 4]
```diff
@@ -1720,7 +1740,9 @@ HapiSpec payTestSelfDestructCall() {
                         cryptoCreate(PAYER).balance(1_000_000_000_000L).logged(),
                         cryptoCreate(RECEIVER).balance(1_000L),
                         uploadInitCode(contract),
-                        contractCreate(contract))
+                    contractCreate(contract),
+                    getContractInfo(contract)
+                        .exposingEvmAddress(cb -> tokenCreateContractAddress.set(asHeadlongAddress(cb))))
```
Comment (vtronkov) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java:1745]: Getting the contract id dynamically because when we convert the test to use eth context it doesn't find the contract address id

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/EthereumCallDataHydration.java [hunk 5]
```diff
@@ -75,9 +75,19 @@ public HydratedEthTxData tryToHydrate(
...
             }
+
+            final var initcodeBytes = callDataFile.contents().toByteArray();
+            // Bytes.fromHex() doesn't appreciate a leading '0x' but we supported it in mono-service
+            final String hexedInitcode;
+            if (initcodeBytes[0] == (byte) '0' && initcodeBytes[1] == (byte) 'x') {
+                hexedInitcode = new String(initcodeBytes, 2, initcodeBytes.length - 2);
+            } else {
+                hexedInitcode = new String(initcodeBytes);
+            }
```
Comment (vtronkov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/EthereumCallDataHydration.java:86]: This snipped is taken from `HevmTransactionFactory.initcodeFor`. The idea is that we have tests the call data is prefixed with `0x`. To be consistent we are removing the prefix.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/HevmTransactionFactory.java [hunk 6]
```diff
@@ -198,6 +198,7 @@ private HederaEvmTransaction fromHapiEthereum(
             @NonNull final AccountID senderId,
             @NonNull final EthTxData ethTxData,
             final long maxGasAllowance) {
+        validateTrue(ethTxData.getAmount() >= 0, CONTRACT_NEGATIVE_VALUE);
```
Comment (vtronkov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/HevmTransactionFactory.java:201]: Missing validation

**Comment by tinker-michaelj** [comment]
Compile error on this branch?

**Comment by vtronkov** [comment]
> Compile error on this branch?

@tinker-michaelj There was this compile error in dev because of a new `ResponseCodeEnum` type and the branch wasn't up-to-date. I updated the branch and now it should be okay.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java [hunk 7]
```diff
@@ -1670,20 +1674,71 @@ public static Tuple nftTransfer(
...
             } else {
                 convertedOps.add(op);
             }
         }
         return convertedOps;
     }
 
+    private static Object[] tryToSwapLongZeroToEVMAddresses(Object[] args, HapiSpec spec) {
+        return Arrays.stream(args)
+                .map(arg -> {
```
Comment (david-bakin-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:1725]: I like using a stream + `map` here but jeez it's kind of ugly.  Can't think of any easy way to clean it up either.  Sigh.

Maybe it would be an improvement to pull lines L1727-L1735 to a separate method, leaving only the test `instanceof Address` inside the `map` ...   ???

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/evm/Evm46ValidationSuite.java [hunk 8]
```diff
@@ -609,6 +613,7 @@ private HapiSpec internalCallToNonExistingMirrorAddressResultsInNoopSuccess() {
                                         resultWith().createdContractIdsCount(0).gasUsed(24972))));
     }
 
+    // Adding refusingEthConversion() due to fee differences and not supported address type
```
Comment (david-bakin-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/evm/Evm46ValidationSuite.java:616]: And again I don't get it.  Could these two questions be simply answered by adding Javadoc over there at `HapiContractCreate.refusingEthConversion` that explains what it does and when you'd want to do it?

(And I see there are multiple others of this comment in this suite and others.  But I won't comment on them; just consider all of them included here.)

**Review by david-bakin-sl — COMMENTED** [review]
Most of these are questions I have or style comments.  I'd like to see the answers to my questions.  But the style comments are suggestions that are up to you.

Let me know when you'd like me to look this over again.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java [hunk 9]
```diff
@@ -1670,20 +1674,71 @@ public static Tuple nftTransfer(
...
+            final String privateKeyRef,
+            final Key adminKey,
+            final long defaultGas,
+            final HapiSpec spec) {
         final var convertedOps = new ArrayList<HapiSpecOperation>(ops.size());
         for (final var op : ops) {
-            if (op instanceof HapiContractCall callOp && callOp.isConvertableToEthCall()) {
+            if (op instanceof HapiContractCall callOp && callOp.isConvertableToEthCall(spec)) {
+                // if we have function params, try to swap the long zero address with the EVM address
+                if (callOp.getParams().length > 0 && callOp.getAbi() != null) {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:1692]: I'm not sure I understand this comment. Do you suggest creating a new class `HapiBaseContractCall` that will sit between `HapiContractCall` and `HapiBaseCall`? So `HapiContractCall` will extend `HapiBaseContractCall` and `HapiBaseContractCall` will extend `HapiBaseCall`? I am also confused about the `Optional` comment for the abi and the params. Currently, they are String and Object[] - do you suggest making them as `Optional`?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/queries/contract/HapiGetContractInfo.java [hunk 10]
```diff
@@ -174,6 +180,12 @@ protected void assertExpectationsGiven(HapiSpec spec) throws Throwable {
         if (registryEntry.isPresent()) {
             spec.registry().saveContractInfo(registryEntry.get(), actualInfo);
         }
+        if (saveEVMAddressToRegistry.isPresent()) {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/queries/contract/HapiGetContractInfo.java:183]: Nice catch 👍 Moved

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnUtils.java [hunk 11]
```diff
@@ -633,4 +636,15 @@ public static boolean isEndOfStakingPeriodRecord(final TransactionRecord record)
     public static boolean isNotEndOfStakingPeriodRecord(final TransactionRecord record) {
         return !isEndOfStakingPeriodRecord(record);
     }
+
+    public static ByteString constructorArgsToByteString(String abi, Object[] args)
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnUtils.java:643]: Added 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnUtils.java [hunk 12]
```diff
@@ -633,4 +636,15 @@ public static boolean isEndOfStakingPeriodRecord(final TransactionRecord record)
...
+            throws UnsupportedEncodingException {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnUtils.java:644]: Looks like the method above(literalInitcodeFor) is wrapping a checked exception in `UncheckedIOException`. I will use the same pattern here.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnVerbs.java [hunk 13]
```diff
@@ -590,6 +590,30 @@ public static HapiSpecOperation uploadInitCode(final Optional<String> payer, fin
...
+    /**
+     *  This method enables uploading a contract bytecode with the constructor parameters (if present) appended at the end of the file
+     *  Used for ethereum create conversion when we need to pass constructor arguments
+     * @param contractName
+     * @param abi
+     * @param args
+     * @return
+     */
+    public static HapiSpecOperation updateInitCodeWithConstructorArgs(
+            final String contractName, final String abi, final Object... args) {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnVerbs.java:626]: Well, the idea was that the `uploadInitCode`'s `payer` is not used anyway so there were no need for us to add it as well. The `uploadInitCode` method is called only once(in `TxnVerbs.uploadInitCode`) and the payer is always `Optional.empty`.

But for the sake of consistency, I added the payer in `updateInitCodeWithConstructorArgs` so we can have the option of passing the payer, if we want.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java [hunk 14]
```diff
@@ -155,7 +155,12 @@ public HederaFunctionality type() {
         return HederaFunctionality.ContractCall;
     }
 
-    public boolean isConvertableToEthCall() {
+    public boolean isConvertableToEthCall(HapiSpec spec) {
+        // check if the key is SECP256K1
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java:159]: Fixed the comment 👍 
About the field name - I think the name should still hold. We haven't changed anything related to that. The only change here is that if the key is `SECP256K1` we always refuse to convert to eth call. This shouldn't change the semantics of the name.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCreate.java [hunk 15]
```diff
@@ -442,4 +446,12 @@ public Optional<Integer> getRetryLimits() {
     public Optional<EnumSet<ResponseCodeEnum>> getPermissiblePrechecks() {
         return permissiblePrechecks;
     }
+
+    public Optional<Long> getFee() {
+        return fee;
+    }
+
+    public void setArgs(Optional<Object[]> args) {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCreate.java:454]: Changed 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumContractCreate.java [hunk 16]
```diff
@@ -143,6 +144,22 @@ public HapiEthereumContractCreate(
...
         this.permissibleStatuses = contractCreate.getPermissibleStatuses();
         this.permissiblePrechecks = contractCreate.getPermissiblePrechecks();
+        this.payer = contractCreate.getPayer();
+        this.fee = contractCreate.getFee();
+
+        // todo check if 5 hbars is proper value - copied from HapiEthereumCall
+        this.maxGasAllowance = Optional.of(FIVE_HBARS);
+    }
+
+    private Optional<BigInteger> weibarsToTinybars(Optional<Long> balance) {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumContractCreate.java:155]: Changed 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumContractCreate.java [hunk 17]
```diff
@@ -264,8 +280,15 @@ protected Consumer<TransactionBody.Builder> opBodyDef(HapiSpec spec) throws Thro
...
-        final byte[] callData =
-                Bytes.fromHexString(new String(fileContents.toByteArray())).toArray();
+        byte[] callData;
+        if (args.isPresent() && abi.isPresent()) {
+            final var bytecode = fileContents.concat(TxnUtils.constructorArgsToByteString(abi.get(), args.get()));
+            callData = Bytes.fromHexString(new String(bytecode.toByteArray())).toArray();
+        } else {
+            callData =
+                    Bytes.fromHexString(new String(fileContents.toByteArray())).toArray();
+        }
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumContractCreate.java:291]: a) works like a charm. b) doesn't work for some reason. I'm not sure why there is a difference 🤷 Will leave it for now as is.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java [hunk 18]
```diff
@@ -1670,20 +1674,71 @@ public static Tuple nftTransfer(
...
-            final List<HapiSpecOperation> ops, final String privateKeyRef, final Key adminKey, final long defaultGas) {
+            final List<HapiSpecOperation> ops,
+            final String privateKeyRef,
+            final Key adminKey,
+            final long defaultGas,
+            final HapiSpec spec) {
         final var convertedOps = new ArrayList<HapiSpecOperation>(ops.size());
         for (final var op : ops) {
-            if (op instanceof HapiContractCall callOp && callOp.isConvertableToEthCall()) {
+            if (op instanceof HapiContractCall callOp && callOp.isConvertableToEthCall(spec)) {
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:1690]: hmm, not sure how to do that since we have multiple conditions in the if statement

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java (hunk 7 above)
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:1731]: Changed 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/records/SnapshotModeOp.java [hunk 19]
```diff
@@ -718,7 +719,12 @@ private void writeSnapshotOf(@NonNull final List<ParsedItem> postPlaceholderItem
         om.writeValue(fout, suiteSnapshots);
     }
 
-    private static Path resourceLocOf(@NonNull final String snapshotLoc, @NonNull final String suiteName) {
+    private static Path resourceLocOf(@NonNull final String snapshotLoc, @NonNull String suiteName) {
+        // If we start a test with Ethereum context we are adding a "_Eth" suffix to test name.
+        // Before we start a test we need to remove this suffix to get the correct snapshot file name.
+        if (suiteName.endsWith(HapiSuite.ETH_SUFFIX)) {
+            suiteName = suiteName.replace(HapiSuite.ETH_SUFFIX, "");
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/records/SnapshotModeOp.java:726]: Yeah, you are correct. If we have more than one `_Eth` this would fail but I think we should be safe here 😄 `String.replaceLast()`  would be really handy here 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java [hunk 20]
```diff
@@ -367,6 +367,15 @@ public static HapiSpecOperation captureChildCreate2MetaFor(
             final var numRecords = response.getChildTransactionRecordsCount();
             int numExpectedChildren = givenNumExpectedChildren;
             int childOfInterest = givenChildOfInterest;
+
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java:370]: I think this applies to almost all methods related to the test-suites. I rarely see comments there and most of the time is hard to understand what's going on.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/evm/Evm46ValidationSuite.java [hunk 21]
```diff
@@ -263,6 +263,8 @@ public List<HapiSpec> getSpecsInSuite() {
                 testBalanceOfForSystemAccounts());
     }
 
+    // Refusing ethereum create conversion, because we get INVALID_SIGNATURE upon tokenAssociate,
+    // since we have CONTRACT_ID key
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/evm/Evm46ValidationSuite.java:267]: I moved the comments near the `refusingEthConversion` so it's easier to understand 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/evm/Evm46ValidationSuite.java (hunk 8 above)
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/evm/Evm46ValidationSuite.java:616]: I started adding all the reasons we might want to refuse eth conversion above the `HapiContractCreate.refusingEthConversion` but realized that we have quite a few. I think if we take this approach it would be confusing to know exactly why a certain method refuses a conversion. That's why I just moved the comments near the usage of `refusingEthConversion` to be clearer.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java [hunk 22]
```diff
@@ -359,4 +364,8 @@ static void doObservedLookup(final HapiSpec spec, final Transaction txn, Consume
     protected MoreObjects.ToStringHelper toStringHelper() {
```
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java]: All I meant here was: having a `toString` helper/builder to make it easier to write your `toString` methods has been part of Guava for a long time _because it is so widely needed_ so why hasn't it been picked up for the JDK itself? (Also for `equals` and `hashcode`.)

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:1692]: No I completely screwed that comment up.  Let me try again.

`HapiContractCall` derives from `HapiBaseCall` which currently has both `params` and `abi` as nullable fields, rather than wrapped in an `Optional`.  

That's different from `HapiContractCreate` which derives from `HapiBaseContractCreate` where all the interesting fields _are_ wrapped in an `Optional`, including `abi` and `args`.

So I'm suggesting fixing up `HapiBaseContractCreate` to use `Optionals` too.  I think the resulting code reads better, when those fields are used using the `Optional` accessors.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java]: ok, that's fine.  just to explain though:  you have a method called `isConvertableToEthCall` which _used to_ just consult the field `convertableToEthCall` and return that field.  So in that earlier code, the method was just an accessor for that field.  So obviously they had the same name.

but _now_ the method actually has to compute something in addition to that field to get to the truth of `isConvertableToEthCall`.  So now it is clear that the _field_ `convertableToEthCall` isn't really specifying that anymore.  It's something different.

And in fact we know what it is, from the setter:  it's a flag indicating you want to _refuse_ Eth conversion.  

So my suggestion was: that's what the field should be named: `refusingEthConversion`.  And then the method `isConvertableToEthCall` lets you convert as long as you're not refusing it explicitly (from considerations elsewhere) or if the key isn't appropriate.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumContractCreate.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiEthereumContractCreate.java]: Wonder why b) didn't work ... I'll have to look at that later to understand `Bytes` better ...

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:1690]: I'm just referring to the _outer_ `if` - I thought Java's (new) switch-pattern-with-guards could do it.  But I didn't try to rewrite it myself.  I'll leave that as an exercise for myself later, to see how it all works ...

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java:370]: Um, yes, so I agree with you.  Um, so, maybe cough up a javadoc for this one method to set an example to all of us for the future???

**Review by david-bakin-sl — DISMISSED** [review]
I made some comments to your comment responses to me, not part of this review (the way github works).

A couple of things that are left for the future: if you can, drop a `// FUTURE:` comment there?

And/or add that javadoc I suggested.

Anyway, besides that: LGTM.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:1692]: Gotcha 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/contract/HapiContractCall.java]: Okay, that makes sense. My problem with that is that if I change the field name from `convertableToEthCall` to `refusingEthConversion` then in `refusingEthConversion` **method** we need to switch the logic and set it up to **true**. Then we will need to switch all the places that use this field and switch the logic. Not that this is a huge problem, but this was my concern.

For now, I extracted the additional logic in a new method `isKeySECP256K1` that is separate from `isConvertableToEthCall`. What do you think about this approach?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java:374]: Okay, I added a comment but since I'm not the one that wrote the method I'm not 100% sure that it's accurate 😄 But it seems like it.

**Review by tinker-michaelj — APPROVED** [review]
LGTM, tyvm @vtronkov !
